libc-top-half/musl/src/string/strspn_simd.c

```diff
@@ -0,0 +1,175 @@
+#if defined(__wasm_simd128__) && defined(__wasilibc_simd_string)
+// Skip Clang 19 and Clang 20 which have a bug (llvm/llvm-project#146574)
+// which results in an ICE when inline assembly is used with a vector result.
+#if __clang_major__ != 19 && __clang_major__ != 20
+
+#include <stdint.h>
+#include <string.h>
+#include <wasm_simd128.h>
+
+#if !defined(__wasm_relaxed_simd__) || !defined(__RELAXED_FN_ATTRS)
+#define wasm_i8x16_relaxed_swizzle wasm_i8x16_swizzle
+#endif
+
+// SIMDized check which bytes are in a set (Geoff Langdale)
+// http://0x80.pl/notesen/2018-10-18-simd-byte-lookup.html
+
+typedef struct {
+  __u8x16 lo;
+  __u8x16 hi;
+} __wasm_v128_bitmap256_t;
+
+__attribute__((always_inline))
+static void __wasm_v128_setbit(__wasm_v128_bitmap256_t *bitmap, int i) {
+  uint8_t hi_nibble = (uint8_t)i >> 4;
+  uint8_t lo_nibble = (uint8_t)i & 0xf;
+  bitmap->lo[lo_nibble] |= (uint8_t)((uint32_t)1 << (hi_nibble - 0));
+  bitmap->hi[lo_nibble] |= (uint8_t)((uint32_t)1 << (hi_nibble - 8));
```

Comment:
I'm interested in understanding the codegen of this: so the `...[lo_nibble] |=` is generating some `i8x16.replace_lane` but somehow also `OR`-ing the high nibble bits? What is emitted by LLVM here?

Reply:
LLVM cheats and uses the stack, `wasm-opt` (and `wasm-ctor-eval`) then removes any traces of `$__stack_pointer` in this build:
https://github.com/ncruces/go-sqlite3/blob/b72fd5db/sqlite3/libc/libc.wat#L1646-L1726